src/target/riscv/riscv-013.c

```diff
@@ -2073,3 +2076,418 @@ void riscv013_clear_abstract_error(struct target *target)
+    riscv_set_register(target, GDB_REGNO_DPC, testvar64);
+    COMPLIANCE_TEST((riscv_get_register(target, GDB_REGNO_DPC) & dpcmask) == (testvar64 & dpcmask), "DPC must be writable.");
+    riscv_set_register(target, GDB_REGNO_DPC, ~testvar64);
+    uint64_t dpc = riscv_get_register(target, GDB_REGNO_DPC);
+    COMPLIANCE_TEST((dpc & dpcmask) == ((~testvar64) & dpcmask), "DPC must be writable");
+    if (hartsel == 0) {bogus_dpc = dpc;} // For a later test step
+  }
+  
+  //NDMRESET
+  // NDMRESET
+  // Asserting non-debug module reset should not reset Debug Module state.
+  // But it should reset Hart State, e.g. DPC should get a different value.
+  // Also make sure that DCSR reports cause of 'HALT' even though previously we single-stepped.
+  
+  // Write some registers. They should not be impacted by ndmreset.
+  dmi_write(target, DMI_COMMAND, 0xFFFFFFFF);
+  
+  for (unsigned int i = 0; i < get_field(abstractcs, DMI_ABSTRACTCS_PROGSIZE); i ++){
+    testvar = (i + 1) * 0x11111111;
+    dmi_write(target, DMI_PROGBUF0 + i, testvar);
+  }
+  
+  for (unsigned int i = 0; i < get_field(abstractcs, DMI_ABSTRACTCS_DATACOUNT); i ++){
+    testvar = (i + 1) * 0x11111111;
+    dmi_write(target, DMI_DATA0 + i, testvar);
+  }
+
+  dmi_write(target, DMI_ABSTRACTAUTO, 0xFFFFFFFF);
+  abstractauto = dmi_read(target, DMI_ABSTRACTAUTO);
+
+  // Pulse reset.
+  
+  target->reset_halt = true;
+  dmcontrol = dmi_read(target, DMI_DMCONTROL);
+  riscv_set_current_hartid(target, 0);
+  assert_reset(target);
+  deassert_reset(target);
+
+  // Verify that most stuff is not affected by ndmreset.
+  COMPLIANCE_TEST(dmi_read(target, DMI_COMMAND) == 0xFFFFFFFF, "NDMRESET should not affect DMI_COMMAND");
+  COMPLIANCE_TEST(get_field(dmi_read(target, DMI_ABSTRACTCS), DMI_ABSTRACTCS_CMDERR)  == CMDERR_NOT_SUPPORTED, "NDMRESET should not affect DMI_ABSTRACTCS");
+  COMPLIANCE_TEST(dmi_read(target, DMI_ABSTRACTAUTO) == abstractauto, "NDMRESET should not affect DMI_ABSTRACTAUTO");
+
+  // Clean up to avoid future test failures
+  dmi_write(target, DMI_ABSTRACTCS, DMI_ABSTRACTCS_CMDERR);
+  dmi_write(target, DMI_ABSTRACTAUTO, 0);
+  
+  for (unsigned int i = 0; i < get_field(abstractcs, DMI_ABSTRACTCS_PROGSIZE); i ++){
+    testvar = (i + 1) * 0x11111111;
+    COMPLIANCE_TEST(dmi_read(target, DMI_PROGBUF0 + i) == testvar, "PROGBUF words must not be affected by NDMRESET");
+  }
+  
+  for (unsigned int i = 0; i < get_field(abstractcs, DMI_ABSTRACTCS_DATACOUNT); i ++){
+    testvar = (i + 1) * 0x11111111;
+    COMPLIANCE_TEST(dmi_read(target, DMI_DATA0 + i) == testvar, "DATA words must not be affected by NDMRESET");
+  }
+
+  // verify that DPC *is* affected by ndmreset. Since we don't know what it *should* be,
+  // just verify that at least it's not the bogus value anymore.
+  COMPLIANCE_TEST(bogus_dpc != 0xdeadbeef, "BOGUS DPC should have been set somehow (bug in compliance test)");
```

Comment:
The low bit of SEPC is always 0.  It's not specified for DPC, but if that works the same way then this test will always pass.

Reply:
This is just trying to make sure you set the `bogus_dpc` variable, it's checking for a bug in the test, not in the HW.